implementations/rust/ockam/ockam_api/src/authenticator/direct/direct_authenticator.rs

```diff
@@ -0,0 +1,155 @@
+
+        if !check.is_enroller {
+            warn!("Not enroller {} is trying to add a member {}", enroller, id);
+            return Ok(Either::Right(DirectAuthenticatorErrorString(
+                "Not enroller is trying to add a member".to_string(),
+            )));
+        }
+
+        let attrs = attrs
+            .iter()
+            .map(|(k, v)| (k.as_bytes().to_vec(), v.as_bytes().to_vec()))
+            .collect();
+
+        // Check if we're trying to create an enroller
+        if EnrollerAccessControlChecks::check_bin_attributes_is_enroller(&attrs) {
+            // Only pre-trusted identities will be able to add enrollers
+            if !check.is_pre_trusted {
+                warn!(
+                    "Not pre trusted enroller {} is trying to create an enroller {}",
+                    enroller, id
+                );
+
+                return Ok(Either::Right(DirectAuthenticatorErrorString(
+                    "Not pre trusted enroller is trying to create an enroller".to_string(),
+                )));
+            }
+        }
+
+        let member = AuthorityMember::new(id.clone(), attrs, Some(enroller.clone()), now()?, false);
+
+        if let Err(err) = self.members.add_member(member).await {
+            warn!("Error adding member {} directly: {}", id, err);
+            return Ok(Either::Right(DirectAuthenticatorErrorString(
+                "Error adding member".to_string(),
+            )));
+        }
+
+        Ok(Either::Left(()))
+    }
+
+    pub async fn list_members(
+        &self,
+        enroller: &Identifier,
+    ) -> Result<Either<HashMap<Identifier, AttributesEntry>, DirectAuthenticatorErrorString>> {
+        let check =
+            EnrollerAccessControlChecks::check_identifier(self.members.clone(), enroller).await?;
+
+        if !check.is_enroller {
+            warn!("Not enroller {} is trying to list members", enroller);
+            return Ok(Either::Right(DirectAuthenticatorErrorString(
+                "Not enroller is trying to list members".to_string(),
+            )));
+        }
+
+        let all_members = self.members.get_members().await?;
+
+        let mut res = HashMap::<Identifier, AttributesEntry>::default();
+        for member in all_members {
+            let entry = AttributesEntry::new(
+                member.attributes().clone(),
+                member.added_at(),
+                None,
+                member.added_by().clone(),
+            );
+            res.insert(member.identifier().clone(), entry);
+        }
+
+        Ok(Either::Left(res))
+    }
+
+    pub async fn delete_member(
+        &self,
+        enroller: &Identifier,
+        identifier: &Identifier,
+    ) -> Result<Either<(), DirectAuthenticatorErrorString>> {
+        let check_enroller =
+            EnrollerAccessControlChecks::check_identifier(self.members.clone(), enroller).await?;
+
+        if !check_enroller.is_enroller {
```

Comment:
This code, up to L.126 could be refactored for the 3 methods.

Reply:
I don't think there is much to refactor here, it's very simple